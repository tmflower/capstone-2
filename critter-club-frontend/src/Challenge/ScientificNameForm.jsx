import { useState, useRef } from 'react';

export function ScientificNameForm({ commonName, scientificName, message, points, setPoints, numQuestions, setNumQuestions  }) {
    
    const initialState = {userGuess: ''}
    const [formData, setFormData] = useState(initialState);
    const [feedback, setFeedback] = useState('');
    const { userGuess } = formData;
    const ref = useRef();

    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct; provide corresponding feedback message
        e.preventDefault();
        if (userGuess.toLowerCase() === scientificName.toLowerCase()) {
            setFeedback(message.correct);
            setPoints(points+=10);
            setNumQuestions(numQuestions+1);
        }
        else {
            setFeedback(message.incorrect);
        }
    }

    const handleReset = (e) => {
        ref.current.value = "";
        setFeedback('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <fieldset>
            <legend>What is the {commonName.toLowerCase()}'s scientific name?</legend>
            <label htmlFor='Scientific name'>Scientific name            
            <input 
                id="userGuess"
                type="text" 
                defaultValue="" 
                ref={ref} 
                name="userGuess"
                onChange={handleChange}
                >
            </input>
            </label>
            <div>
                { !feedback ? 
                <button type="submit">Check answer</button>
                : null }
                { feedback === message.incorrect ?
                <button type="reset" onClick={handleReset}>Try again</button>
                : null }
            </div>
            <p>{feedback}</p>
            </fieldset>
        </form>
    )
}