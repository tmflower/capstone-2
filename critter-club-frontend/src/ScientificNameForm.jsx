import { useState } from 'react';

export function ScientificNameForm({commonName, scientificName}) {

    const [formData, setFormData] = useState('');
    const { userGuess } = formData;

    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct
        // TODO: add functionality for user feedback
        e.preventDefault();
        if (userGuess === scientificName) {
            console.log("Correct on scientific name!")
        }
        else console.log("sorry, try again");
    }

    const handleReset = (e) => {
        setFormData('');
    }

    return (
        <form onSubmit={handleSubmit}>
            <fieldset>
            <legend>What is the {commonName.toLowerCase()}'s scientific name?</legend>
            <label>Scientific name
            
            <input 
                type="text"
                name="userGuess"
                onChange={handleChange}
                >
            </input>
            </label>
            <div>
                <button type="submit">Check answer</button>
                <button type="reset" onClick={handleReset}>Clear answer</button>
            </div>
            </fieldset>
        </form>
    )
}