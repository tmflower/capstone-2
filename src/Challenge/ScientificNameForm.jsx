import { useState, useRef } from 'react';
import { Paper, FormLabel, Typography, Button } from '@mui/material';
import { theme } from '../theme/theme';

/** ScientificNameForm renders a question with a text input */

export function ScientificNameForm({ 
    commonName, 
    scientificName, 
    message, 
    points, 
    setPoints, 
    numQuestions, 
    setNumQuestions  }) {
    
    // Display form with no default answers and no initial message
    const initialState = {userGuess: ''}
    const [formData, setFormData] = useState(initialState);
    const [feedback, setFeedback] = useState('');
    const ref = useRef();

    // Deconstruct formData properties
    const { userGuess } = formData;
    

    // Set formData to user's input
    const handleChange = (e) => {
        const { name, value } = e.target;
		setFormData((formData) => 
            ({ ...formData, [name]: value }));
    }

    // compare user's selection to animal data to check if correct; provide corresponding feedback message
    // update user's points
    // increment numQuestions; when numQuestions is >=3, user can submit answers and earn badge
    const handleSubmit = (e) => {
        // compare user's selection to animal data to check if correct; provide corresponding feedback message
        e.preventDefault();
        if (userGuess.toLowerCase() === scientificName.toLowerCase()) {
            setFeedback(message.correct);
            setPoints(points+=10);
            setNumQuestions(numQuestions+1);
        }
        else {
            setFeedback(message.incorrect);
        }
    }

    // clear form responses & message when user clicks "Try again" button
    const handleReset = (e) => {
        ref.current.value = "";
        setFeedback('');
    }

    return (
        <Paper
        elevation={8}
        sx={{ padding: 3, mt: 3 }}>
        <form onSubmit={handleSubmit}>
            <fieldset>
            <FormLabel sx={{ fontSize: '1.5rem', fontFamily: 'Lexend Deca, Arial',}}>
                <legend>What is the {commonName.toLowerCase()}'s scientific name?</legend>
            </FormLabel>
            <label htmlFor='Scientific name'>Scientific name            
            <input                 
                type="text" 
                name="userGuess"
                id="Scientific name"
                defaultValue="" 
                className="input-nonMUI"
                ref={ref}                 
                onChange={handleChange}                
            />
            </label>
            <div>
                { !feedback ? 
                    <Button type="submit">Check answer</Button>
                : null }
                { feedback === message.incorrect ?
                    <Button 
                        id="alt-button" 
                        type="reset" 
                        onClick={handleReset}>Clear answer
                    </Button>
                    : null }
            </div>
            <Typography 
                id="quiz-feedback" 
                variant="h5" 
                sx={{ fontFamily: theme.typography.primary, textAlign: 'center' }}>{feedback}
            </Typography>
            </fieldset>
        </form>
        </Paper>
    )
}