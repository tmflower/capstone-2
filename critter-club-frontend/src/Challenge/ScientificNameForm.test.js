import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { ScientificNameForm } from './ScientificNameForm';
import UserContext from '../userContext';
import userEvent from "@testing-library/user-event";


const currentUser = {user: { username: "Tiny", id: 1, password: 'abc123$%^', points: 40, userBadges: ['Koala', 'Jellyfish', 'Painted Turtle'], parentId: 1 }}

const message = {
  correct: "⭐️⭐️⭐️ You got it! ⭐️⭐️⭐️",
  incorrect: "😧😧😧 Sorry, try again. 😧😧😧"
};

// smoke test
test("renders without crashing", () => {
    render(
    <MemoryRouter>
      <UserContext.Provider value={currentUser}>
        <ScientificNameForm
          commonName="Hippopotamus"
          message={message} />
      </UserContext.Provider>      
    </MemoryRouter>);
  });

// snapshot test
test("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <UserContext.Provider value={currentUser}>
          <ScientificNameForm
            commonName="Hippopotamus"
            message={message} />
        </UserContext.Provider>        
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});