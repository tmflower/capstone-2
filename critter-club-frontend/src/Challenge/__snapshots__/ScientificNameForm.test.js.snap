// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-2e4xct-MuiPaper-root"
  >
    <form>
      <fieldset>
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-tshmje-MuiFormLabel-root"
        >
          <legend>
            What is the hippopotamus's scientific name?
          </legend>
        </label>
        <label
          for="Scientific name"
        >
          Scientific name
          <input
            class="input-nonMUI"
            id="Scientific name"
            name="userGuess"
            type="text"
            value=""
          />
        </label>
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Check answer
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-18czui-MuiTypography-root"
          id="quiz-feedback"
        />
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
